diff --git a/scene/2d/physics_body_2d.cpp b/scene/2d/physics_body_2d.cpp
index d42bd6adaf..8ee24bf13b 100644
--- a/scene/2d/physics_body_2d.cpp
+++ b/scene/2d/physics_body_2d.cpp
@@ -1266,6 +1266,8 @@ Vector2 KinematicBody2D::move_and_slide(const Vector2 &p_linear_velocity, const
 				if (p_up_direction == Vector2()) {
 					//all is a wall
 					on_wall = true;
+					motion = slide_wall(motion, p_up_direction, p_floor_max_angle, collision);
+					body_velocity = slide_wall(body_velocity, p_up_direction, p_floor_max_angle, collision);
 				} else {
 					if (Math::acos(collision.normal.dot(p_up_direction)) <= p_floor_max_angle + FLOOR_ANGLE_THRESHOLD) { //floor
 
@@ -1282,15 +1284,20 @@ Vector2 KinematicBody2D::move_and_slide(const Vector2 &p_linear_velocity, const
 								return Vector2();
 							}
 						}
+
+						motion = slide_floor(motion, p_up_direction, p_floor_max_angle, collision);
+						body_velocity = slide_floor(body_velocity, p_up_direction, p_floor_max_angle, collision);
+
 					} else if (Math::acos(collision.normal.dot(-p_up_direction)) <= p_floor_max_angle + FLOOR_ANGLE_THRESHOLD) { //ceiling
 						on_ceiling = true;
+						motion = slide_ceiling(motion, p_up_direction, p_floor_max_angle, collision);
+						body_velocity = slide_ceiling(body_velocity, p_up_direction, p_floor_max_angle, collision);
 					} else {
 						on_wall = true;
+						motion = slide_wall(motion, p_up_direction, p_floor_max_angle, collision);
+						body_velocity = slide_wall(body_velocity, p_up_direction, p_floor_max_angle, collision);
 					}
 				}
-
-				motion = motion.slide(collision.normal);
-				body_velocity = body_velocity.slide(collision.normal);
 			}
 		}
 
@@ -1303,6 +1310,51 @@ Vector2 KinematicBody2D::move_and_slide(const Vector2 &p_linear_velocity, const
 	return body_velocity;
 }
 
+Vector2 KinematicBody2D::slide_floor(Vector2 p_motion, Vector2 p_up_direction, float p_floor_max_angle, Collision p_collision) {
+	if (get_script_instance() && get_script_instance()->has_method("slide_floor")) {
+		if (motion_cache.is_null()) {
+			motion_cache.instance();
+			motion_cache->owner = this;
+		}
+
+		motion_cache->collision = p_collision;
+
+		return get_script_instance()->call("slide_floor", p_motion, p_up_direction, p_floor_max_angle, motion_cache);
+	} else {
+		return p_motion.slide(p_collision.normal);
+	}
+}
+
+Vector2 KinematicBody2D::slide_ceiling(Vector2 p_motion, Vector2 p_up_direction, float p_floor_max_angle, Collision p_collision) {
+	if (get_script_instance() && get_script_instance()->has_method("slide_ceiling")) {
+		if (motion_cache.is_null()) {
+			motion_cache.instance();
+			motion_cache->owner = this;
+		}
+
+		motion_cache->collision = p_collision;
+
+		return get_script_instance()->call("slide_ceiling", p_motion, p_up_direction, p_floor_max_angle, motion_cache);
+	} else {
+		return p_motion.slide(p_collision.normal);
+	}
+}
+
+Vector2 KinematicBody2D::slide_wall(Vector2 p_motion, Vector2 p_up_direction, float p_floor_max_angle, Collision p_collision) {
+	if (get_script_instance() && get_script_instance()->has_method("slide_wall")) {
+		if (motion_cache.is_null()) {
+			motion_cache.instance();
+			motion_cache->owner = this;
+		}
+
+		motion_cache->collision = p_collision;
+
+		return get_script_instance()->call("slide_wall", p_motion, p_up_direction, p_floor_max_angle, motion_cache);
+	} else {
+		return p_motion.slide(p_collision.normal);
+	}
+}
+
 Vector2 KinematicBody2D::move_and_slide_with_snap(const Vector2 &p_linear_velocity, const Vector2 &p_snap, const Vector2 &p_up_direction, bool p_stop_on_slope, int p_max_slides, float p_floor_max_angle, bool p_infinite_inertia) {
 
 	bool was_on_floor = on_floor;
@@ -1495,6 +1547,10 @@ void KinematicBody2D::_bind_methods() {
 
 	ClassDB::bind_method(D_METHOD("_direct_state_changed"), &KinematicBody2D::_direct_state_changed);
 
+	BIND_VMETHOD(MethodInfo(Variant::VECTOR2, "slide_floor", PropertyInfo(Variant::VECTOR2, "motion"), PropertyInfo(Variant::VECTOR2, "up_direction"), PropertyInfo(Variant::REAL, "floor_max_angle"), PropertyInfo(Variant::OBJECT, "collision", PROPERTY_HINT_RESOURCE_TYPE, "KinematicCollision2D")));
+	BIND_VMETHOD(MethodInfo(Variant::VECTOR2, "slide_ceiling", PropertyInfo(Variant::VECTOR2, "motion"), PropertyInfo(Variant::VECTOR2, "up_direction"), PropertyInfo(Variant::REAL, "floor_max_angle"), PropertyInfo(Variant::OBJECT, "collision", PROPERTY_HINT_RESOURCE_TYPE, "KinematicCollision2D")));
+	BIND_VMETHOD(MethodInfo(Variant::VECTOR2, "slide_wall", PropertyInfo(Variant::VECTOR2, "motion"), PropertyInfo(Variant::VECTOR2, "up_direction"), PropertyInfo(Variant::REAL, "floor_max_angle"), PropertyInfo(Variant::OBJECT, "collision", PROPERTY_HINT_RESOURCE_TYPE, "KinematicCollision2D")));
+
 	ADD_PROPERTY(PropertyInfo(Variant::REAL, "collision/safe_margin", PROPERTY_HINT_RANGE, "0.001,256,0.001"), "set_safe_margin", "get_safe_margin");
 	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "motion/sync_to_physics"), "set_sync_to_physics", "is_sync_to_physics_enabled");
 }
diff --git a/scene/2d/physics_body_2d.h b/scene/2d/physics_body_2d.h
index 6766bafde3..7ce441d214 100644
--- a/scene/2d/physics_body_2d.h
+++ b/scene/2d/physics_body_2d.h
@@ -192,6 +192,10 @@ private:
 	void _body_inout(int p_status, ObjectID p_instance, int p_body_shape, int p_local_shape);
 	void _direct_state_changed(Object *p_state);
 
+	Vector2 slide_floor(Vector2 p_motion, Vector2 p_up_direction, float p_floor_max_angle, Collision p_collision);
+	Vector2 slide_ceiling(Vector2 p_motion, Vector2 p_up_direction, float p_floor_max_angle, Collision p_collision);
+	Vector2 slide_wall(Vector2 p_motion, Vector2 p_up_direction, float p_floor_max_angle, Collision p_collision); 
+
 	bool _test_motion(const Vector2 &p_motion, bool p_infinite_inertia = true, float p_margin = 0.08, const Ref<Physics2DTestMotionResult> &p_result = Ref<Physics2DTestMotionResult>());
 
 protected:
diff --git a/scene/3d/physics_body.cpp b/scene/3d/physics_body.cpp
index caeae90238..2453580742 100644
--- a/scene/3d/physics_body.cpp
+++ b/scene/3d/physics_body.cpp
@@ -1191,6 +1191,8 @@ Vector3 KinematicBody::move_and_slide(const Vector3 &p_linear_velocity, const Ve
 				if (p_up_direction == Vector3()) {
 					//all is a wall
 					on_wall = true;
+					motion = slide_wall(motion, p_up_direction, p_floor_max_angle, collision);
+					body_velocity = slide_wall(body_velocity, p_up_direction, p_floor_max_angle, collision);
 				} else {
 					if (Math::acos(collision.normal.dot(p_up_direction)) <= p_floor_max_angle + FLOOR_ANGLE_THRESHOLD) { //floor
 
@@ -1207,16 +1209,19 @@ Vector3 KinematicBody::move_and_slide(const Vector3 &p_linear_velocity, const Ve
 								return Vector3();
 							}
 						}
+						motion = slide_floor(motion, p_up_direction, p_floor_max_angle, collision);
+						body_velocity = slide_floor(body_velocity, p_up_direction, p_floor_max_angle, collision);
 					} else if (Math::acos(collision.normal.dot(-p_up_direction)) <= p_floor_max_angle + FLOOR_ANGLE_THRESHOLD) { //ceiling
 						on_ceiling = true;
+						motion = slide_ceiling(motion, p_up_direction, p_floor_max_angle, collision);
+						body_velocity = slide_ceiling(body_velocity, p_up_direction, p_floor_max_angle, collision);
 					} else {
 						on_wall = true;
+						motion = slide_wall(motion, p_up_direction, p_floor_max_angle, collision);
+						body_velocity = slide_wall(body_velocity, p_up_direction, p_floor_max_angle, collision);
 					}
 				}
 
-				motion = motion.slide(collision.normal);
-				body_velocity = body_velocity.slide(collision.normal);
-
 				for (int j = 0; j < 3; j++) {
 					if (locked_axis & (1 << j)) {
 						body_velocity[j] = 0;
@@ -1234,6 +1239,53 @@ Vector3 KinematicBody::move_and_slide(const Vector3 &p_linear_velocity, const Ve
 	return body_velocity;
 }
 
+Vector3 KinematicBody::slide_floor(Vector3 p_motion, Vector3 p_up_direction, float p_floor_max_angle, Collision p_collision) {
+	if (get_script_instance() && get_script_instance()->has_method("slide_floor")) {
+		if (motion_cache.is_null()) {
+			motion_cache.instance();
+			motion_cache->owner = this;
+		}
+
+		motion_cache->collision = p_collision;
+
+		return get_script_instance()->call("slide_floor", p_motion, p_up_direction, p_floor_max_angle, motion_cache);
+	} else {
+		return p_motion.slide(p_collision.normal);
+	}
+}
+
+Vector3 KinematicBody::slide_ceiling(Vector3 p_motion, Vector3 p_up_direction, float p_floor_max_angle, Collision p_collision) {
+	if (get_script_instance() && get_script_instance()->has_method("slide_ceiling")) {
+		if (motion_cache.is_null()) {
+			motion_cache.instance();
+			motion_cache->owner = this;
+		}
+
+		motion_cache->collision = p_collision;
+
+		return get_script_instance()->call("slide_ceiling", p_motion, p_up_direction, p_floor_max_angle, motion_cache);
+	} else {
+		return p_motion.slide(p_collision.normal);
+	}
+}
+
+Vector3 KinematicBody::slide_wall(Vector3 p_motion, Vector3 p_up_direction, float p_floor_max_angle, Collision p_collision) {
+	if (get_script_instance() && get_script_instance()->has_method("slide_wall")) {
+		if (motion_cache.is_null()) {
+			motion_cache.instance();
+			motion_cache->owner = this;
+		}
+
+		motion_cache->collision = p_collision;
+
+		return get_script_instance()->call("slide_wall", p_motion, p_up_direction, p_floor_max_angle, motion_cache);
+	} else {
+		return p_motion.slide(p_collision.normal);
+	}
+}
+
+
+
 Vector3 KinematicBody::move_and_slide_with_snap(const Vector3 &p_linear_velocity, const Vector3 &p_snap, const Vector3 &p_up_direction, bool p_stop_on_slope, int p_max_slides, float p_floor_max_angle, bool p_infinite_inertia) {
 
 	bool was_on_floor = on_floor;
@@ -1420,6 +1472,10 @@ void KinematicBody::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("get_slide_count"), &KinematicBody::get_slide_count);
 	ClassDB::bind_method(D_METHOD("get_slide_collision", "slide_idx"), &KinematicBody::_get_slide_collision);
 
+	BIND_VMETHOD(MethodInfo(Variant::VECTOR3, "slide_floor", PropertyInfo(Variant::VECTOR3, "motion"), PropertyInfo(Variant::VECTOR3, "up_direction"), PropertyInfo(Variant::REAL, "floor_max_angle"), PropertyInfo(Variant::OBJECT, "collision", PROPERTY_HINT_RESOURCE_TYPE, "KinematicCollision")));
+	BIND_VMETHOD(MethodInfo(Variant::VECTOR3, "slide_ceiling", PropertyInfo(Variant::VECTOR3, "motion"), PropertyInfo(Variant::VECTOR3, "up_direction"), PropertyInfo(Variant::REAL, "floor_max_angle"), PropertyInfo(Variant::OBJECT, "collision", PROPERTY_HINT_RESOURCE_TYPE, "KinematicCollision")));
+	BIND_VMETHOD(MethodInfo(Variant::VECTOR3, "slide_wall", PropertyInfo(Variant::VECTOR3, "motion"), PropertyInfo(Variant::VECTOR3, "up_direction"), PropertyInfo(Variant::REAL, "floor_max_angle"), PropertyInfo(Variant::OBJECT, "collision", PROPERTY_HINT_RESOURCE_TYPE, "KinematicCollision")));
+
 	ADD_PROPERTYI(PropertyInfo(Variant::BOOL, "move_lock_x", PROPERTY_HINT_NONE, "", PROPERTY_USAGE_NOEDITOR), "set_axis_lock", "get_axis_lock", PhysicsServer::BODY_AXIS_LINEAR_X);
 	ADD_PROPERTYI(PropertyInfo(Variant::BOOL, "move_lock_y", PROPERTY_HINT_NONE, "", PROPERTY_USAGE_NOEDITOR), "set_axis_lock", "get_axis_lock", PhysicsServer::BODY_AXIS_LINEAR_Y);
 	ADD_PROPERTYI(PropertyInfo(Variant::BOOL, "move_lock_z", PROPERTY_HINT_NONE, "", PROPERTY_USAGE_NOEDITOR), "set_axis_lock", "get_axis_lock", PhysicsServer::BODY_AXIS_LINEAR_Z);
diff --git a/scene/3d/physics_body.h b/scene/3d/physics_body.h
index 05bcbe22f0..714a58c2bc 100644
--- a/scene/3d/physics_body.h
+++ b/scene/3d/physics_body.h
@@ -315,6 +315,10 @@ private:
 	Ref<KinematicCollision> _move(const Vector3 &p_motion, bool p_infinite_inertia = true, bool p_exclude_raycast_shapes = true, bool p_test_only = false);
 	Ref<KinematicCollision> _get_slide_collision(int p_bounce);
 
+	Vector3 slide_floor(Vector3 p_motion, Vector3 p_up_direction, float p_floor_max_angle, Collision p_collision);
+	Vector3 slide_ceiling(Vector3 p_motion, Vector3 p_up_direction, float p_floor_max_angle, Collision p_collision);
+	Vector3 slide_wall(Vector3 p_motion, Vector3 p_up_direction, float p_floor_max_angle, Collision p_collision);
+
 protected:
 	void _notification(int p_what);
 	static void _bind_methods();
